### PR TITLE
Don't require Origin header for JSON requests

### LINKED ISSUE #659: Do NOT require Origin header for json requests
currently, an Origin header is always required for POSTs.

if the Accept header is explicitly `Accept:application/json` or the `produces` property favors json, the `Origin` header should not be required.

### PR DESCRIPTION
**Pull request by mraible**
Fixes #659.

### CODE REVIEW

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/filter/oauth/OriginAccessTokenRequestAuthorizer.java [hunk 1]
```diff
@@ -73,61 +83,68 @@ public ServerUriResolver getServerUriResolver() {
     @Override
     public void assertAuthorized(HttpServletRequest request, HttpServletResponse response) throws OAuthException {
 
-        boolean localhostClient = isLocalhostClient(request, response);
+        // #659: if the Accept header is explicitly Accept:application/json or the
+        // produces property favors json, the Origin header should not be required.
+        String accepts = Strings.clean(request.getHeader(ACCEPTS_HEADER_NAME));
+        boolean originHeaderNotRequired = accepts.contains(MediaType.APPLICATION_JSON_VALUE) || producesFavorsJSON;
+
+        if (!originHeaderNotRequired) {
```
Comment (dogeared) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/filter/oauth/OriginAccessTokenRequestAuthorizer.java]: nit (not a showstopper): Could this be made to be `originHeaderRequired` so the `if` is not a double negative?

**Comment by dogeared** [comment]
This looks good. I am UAT'ing against the TCK and will report back here.

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/filter/oauth/OriginAccessTokenRequestAuthorizer.java (hunk 1 above)
Comment (mraible) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/filter/oauth/OriginAccessTokenRequestAuthorizer.java]: Yes, this is possible, but then the logic to create the variable has two double negatives.

```java
boolean originHeaderRequired = !accepts.contains(MediaType.APPLICATION_JSON_VALUE) || !producesFavorsJSON;
```

**Comment by dogeared** [comment]
UAT looks good. Please merge after CI.

FWIW, I tested by removing all the instances of `.header("Origin", webappurl)` in the `Oauth2IT` test of the TCK and all the tests that were passing, still are. This proves that the presence of the `Origin` header is not required.

**Comment by mraible** [comment]
@dogeared Did you see @lhazlewood's comments on #659?

**Comment by dogeared** [comment]
@mraible @lhazlewood - based on our conversation at today's SDK planning meeting, we will proceed with this. However, it is currently failing UAT for servlet, Spring WebMVC, Spring Security Spring WebMVC, Spring Boot WebMVC and Spring Security Spring Boot WebMVC:

*PASS*: Should not return an origin error:
```
curl -X POST --data 'grant_type=password&username=foo&password=bar' \
         -H 'Origin: http://localhost:8080' http://localhost:8080/oauth/token
```


*FAIL*: Should return an origin error:
```
curl -X POST --data 'grant_type=password&username=foo&password=bar' \
        http://localhost:8080/oauth/token
```

(It is still returning a json response with access token and refresh token)

*PASS8: Should not return an origin error:
```
curl -X POST --data 'grant_type=password&username=foo&password=bar' \
        -H 'Accept: application/json' http://localhost:8080/oauth/token
```

**Comment by mraible** [comment]
The reason the following doesn't produce an origin error is because we favor JSON by default:

```
curl -X POST --data 'grant_type=password&username=foo&password=bar' \
        http://localhost:8080/oauth/token
```

Here's my UAT findings with the default configuration in examples/servlet:

**PASS**: Should not return an origin error:
```
curl -X POST --data 'grant_type=password&username=foo&password=bar' \
         -H 'Origin: http://localhost:8080' http://localhost:8080/oauth/token | jq
```

Response:
```json
{
  "error": "invalid_request",
  "error_description": "Unable to authenticate access token request: ",
  "error_uri": "Login attempt failed because there is no Account in the Application's associated Account Stores with the specified username or email."
}
```

**FAIL**: Should not return an origin error (because we favor JSON by default):
```
curl -X POST --data 'grant_type=password&username=foo&password=bar' \
        http://localhost:8080/oauth/token | jq
```

Response:
```json
{
  "error": "invalid_request",
  "error_description": "Unable to authenticate access token request: ",
  "error_uri": "Login attempt failed because there is no Account in the Application's associated Account Stores with the specified username or email."
}
```

**PASS**: Should not return an origin error:
```
curl -X POST --data 'grant_type=password&username=foo&password=bar' \
        -H 'Accept: application/json' http://localhost:8080/oauth/token | jq
```

Response:
```json
{
  "error": "invalid_request",
  "error_description": "Unable to authenticate access token request: ",
  "error_uri": "Login attempt failed because there is no Account in the Application's associated Account Stores with the specified username or email."
}
```

To change the configuration so we don't favor JSON, I created `examples/servlet/src/main/resources/stormpath.properties` with the following:

```
stormpath.web.produces = text/html,application/json
```

Then I restarted Tomcat and tried all the scenarios again:

**PASS**: Should not return an origin error (because we're passing in an Origin header):
```
curl -X POST --data 'grant_type=password&username=foo&password=bar' \
         -H 'Origin: http://localhost:8080' http://localhost:8080/oauth/token | jq
```

Response:
```json
{
  "error": "invalid_request",
  "error_description": "Unable to authenticate access token request: ",
  "error_uri": "Login attempt failed because there is no Account in the Application's associated Account Stores with the specified username or email."
}
```

**FAIL**: Should return an origin error (because we favor HTML by default):
```
curl -X POST --data 'grant_type=password&username=foo&password=bar' \
        http://localhost:8080/oauth/token | jq
```

Response:
```json
{
  "error": "invalid_client",
  "error_description": "Missing Origin or Referer header (Origin preferred)."
}
```

**PASS**: Should not return an origin error (because we pass in Accept JSON header):
```
curl -X POST --data 'grant_type=password&username=foo&password=bar' \
        -H 'Accept: application/json' http://localhost:8080/oauth/token | jq
```

Response:
```json
{
  "error": "invalid_request",
  "error_description": "Unable to authenticate access token request: ",
  "error_uri": "Login attempt failed because there is no Account in the Application's associated Account Stores with the specified username or email."
}
```

**Comment by mraible** [comment]
Back to you @dogeared.

**Comment by dogeared** [comment]
LGTM! Please merge 👍

**Pull request merged**